File: client/lib/shared/models/currency.dart
import 'package:finvo/i18n/strings.g.dart';

/// Supported currencies for the application
/// G9 countries + TWD + HKD
enum Currency {
  // G9 Countries & Most Common
  usd('USD', 'US Dollar', '\$', '🇺🇸', 2),
  cny('CNY', 'Chinese Yuan', '¥', '🇨🇳', 2),
  eur('EUR', 'Euro', '€', '🇪🇺', 2),
  gbp('GBP', 'British Pound', '£', '🇬🇧', 2),
  jpy('JPY', 'Japanese Yen', '¥', '🇯🇵', 0),
  cad('CAD', 'Canadian Dollar', 'C\$', '🇨🇦', 2),
  aud('AUD', 'Australian Dollar', 'A\$', '🇦🇺', 2),
  inr('INR', 'Indian Rupee', '₹', '🇮🇳', 2),
  rub('RUB', 'Russian Ruble', '₽', '🇷🇺', 2),

  // Additional currencies
  hkd('HKD', 'Hong Kong Dollar', 'HK\$', '🇭🇰', 2),
  twd('TWD', 'New Taiwan Dollar', 'NT\$', '🇹🇼', 2);

  final String code;
  final String name;
  final String symbol;
  final String flag;

  /// Fraction digits conventionally displayed for this currency.
  /// Zero-decimal currencies (e.g. JPY) must not render ".00".
  final int decimalDigits;

  const Currency(
    this.code,
    this.name,
    this.symbol,
    this.flag,
    this.decimalDigits,
  );

  /// Default code, as a const for default parameter values and @Default.
  static const String defaultCode = 'CNY';

  /// Get currency by code
  static Currency? fromCode(String code) {
    final upperCode = code.toUpperCase();
    final index = Currency.values.indexWhere(
      (c) => c.code.toUpperCase() == upperCode,
    );
    return index == -1 ? null : Currency.values[index];
  }

  /// Get display name with symbol
  String get displayName => '$localizedName ($symbol)';

  /// Get display name with flag
  String get displayNameWithFlag => '$flag $code - $localizedName';

  /// Get localized name using i18n
  String get localizedName {
    switch (this) {
      case Currency.usd:
        return t.currency.usd;
      case Currency.cny:
        return t.currency.cny;
      case Currency.eur:
        return t.currency.eur;
      case Currency.gbp:
        return t.currency.gbp;
      case Currency.jpy:
        return t.currency.jpy;
      case Currency.cad:
        return t.currency.cad;
      case Currency.aud:
        return t.currency.aud;
      case Currency.inr:
        return t.currency.inr;
      case Currency.rub:
        return t.currency.rub;
      case Currency.hkd:
        return t.currency.hkd;
      case Currency.twd:
        return t.currency.twd;
    }
  }
}

File: server/app/core/constants/currency.py
"""Centralized currency default constants.

Single Source of Truth for all currency-related defaults across the project.
All models, schemas, and inference logic MUST import from here instead of
hardcoding currency strings.

Design decisions:
- PROJECT_DEFAULT_CURRENCY = "CNY" because the product primarily targets
  Chinese-speaking users (base_locale=zh, DB server_defaults are CNY).
- Currency inference runs ONLY ONCE at registration time (via infer_currency).
  After that, the user's primary_currency is immutable unless explicitly
  changed through the settings API. There is NO background job or login hook
  that re-infers currency from locale.
"""

# The project-wide default currency used as:
# 1. SQLAlchemy model column default (Python-side)
# 2. Fallback when currency inference has no signal (no locale, no timezone)
# 3. UPSERT fallback in user_service when creating settings on-the-fly
PROJECT_DEFAULT_CURRENCY: str = "CNY"

# Rich metadata for every supported currency. This is THE single source of
# truth for the currency list; both display output (symbols/names) and
# inference membership derive from it (see currency_utils / currency_inference).
SUPPORTED_CURRENCIES: list[dict[str, str]] = [
    {"code": "USD", "name": "US Dollar", "symbol": "$", "flag": "🇺🇸"},
    {"code": "CNY", "name": "Chinese Yuan", "symbol": "¥", "flag": "🇨🇳"},
    {"code": "EUR", "name": "Euro", "symbol": "€", "flag": "🇪🇺"},
    {"code": "GBP", "name": "British Pound", "symbol": "£", "flag": "🇬🇧"},
    {"code": "JPY", "name": "Japanese Yen", "symbol": "¥", "flag": "🇯🇵"},
    {"code": "CAD", "name": "Canadian Dollar", "symbol": "C$", "flag": "🇨🇦"},
    {"code": "AUD", "name": "Australian Dollar", "symbol": "A$", "flag": "🇦🇺"},
    {"code": "INR", "name": "Indian Rupee", "symbol": "₹", "flag": "🇮🇳"},
    {"code": "RUB", "name": "Russian Ruble", "symbol": "₽", "flag": "🇷🇺"},
    {"code": "HKD", "name": "Hong Kong Dollar", "symbol": "HK$", "flag": "🇭🇰"},
    {"code": "TWD", "name": "New Taiwan Dollar", "symbol": "NT$", "flag": "🇹🇼"},
]

# Canonical set of supported currency codes (derived, used for membership checks).
SUPPORTED_CURRENCY_CODES: frozenset[str] = frozenset(c["code"] for c in SUPPORTED_CURRENCIES)

# Code -> display symbol (derived).
CURRENCY_SYMBOLS: dict[str, str] = {c["code"]: c["symbol"] for c in SUPPORTED_CURRENCIES}
